operatorcourier/flatten.py

```diff
@@ -0,0 +1,151 @@
+import logging
+import os
+from yaml import safe_load, MarkedYAMLError
+import semver
+import ntpath
+import operatorcourier.identify as identify
+import operatorcourier.errors as errors
+
+logger = logging.getLogger(__name__)
+
+
+def get_flattened_files_info(source_dir: str) -> [(str, str)]:
+    """
+    :param source_dir: Path of the directory containing different versions
+    of operator bundles (CRD, CSV, package) in separate version directories
+    :return: A list of tuples where in each tuple, the first element is
+    the source file path to be copied to the flattened directory, and the second is
+    the new file name to be used in the file copy. The function returns an empty list if
+    the source_dir is already flat
+    """
+
+    # extract package file and version folders from source_dir
+    _, folder_names, file_names = next(os.walk(source_dir))
+    if len(folder_names) == 0:
+        logger.info('The source directory is already flat. Exiting the program...')
+        return []
+
+    file_paths_to_copy = []  # [ (SRC_FILE_PATH, NEW_FILE_NAME) ]
+
+    crd_dict = {}  # { CRD_NAME => (VERSION, CRD_PATH) }
+    csv_paths = []
+
+    for version_folder_name in folder_names:
+        parse_version_folder(source_dir, version_folder_name, csv_paths, crd_dict)
+
+    # 1. add package in source_dir
+    package_path, package_name = get_package_file(source_dir, file_names)
+    file_paths_to_copy.append((package_path, package_name))
+
+    # 2. add all CRDs with the latest version
+    for crd_name, (_, crd_path) in crd_dict.items():
+        file_paths_to_copy.append((crd_path, ntpath.basename(crd_path)))
+
+    # 3. add all CSVs
+    for csv_file_name, csv_entries in create_csv_dict(csv_paths).items():
+        if len(csv_entries) == 1:
+            file_paths_to_copy.append((csv_entries[0][1], csv_file_name))
+            continue
+        # duplicate CSVs found in different versions
+        for (version, csv_path) in csv_entries:
+            basename, ext = os.path.splitext(csv_file_name)
+            file_paths_to_copy.append((csv_path, f'{basename}_v{version}{ext}'))
```

Comment:
```suggestion
            file_paths_to_copy.append((csv_path, f'{basename}.v{version}{ext}'))
```

Would it be more consistent to use `.` as a separator here?

Reply:
I would actually prefer ` - ` to ` . ` or ` _ `, since the underscore character can be problematic and the period character is used in the version string itself.

Reply:
Is this matching the filename of the CSV or the CSV object name?
Also, should we be pushing our own opinions on the naming conventions or following existing conventions set by the operator-sdk? Or be flexible?
I ask because there are operators today that follow different conventions, ie. with and without the v, periods as separators, dashes, etc.

Reply:
Literally just the filename of the CSV. I think this problem you are describing @SamiSousa is orthogonal to the issue we are trying to solve. We are not suggesting or being opinionated at all about what you *should* name your files. We are just renaming them to avoid accidental collisions before we push. These files we create shouldn't be checked in or used for any other purpose, just used as an intermediary step at runtime.